cpu/sam0_common/sam0_sdhc/sdhc.c

```diff
@@ -319,7 +319,7 @@ int sdhc_init(sdhc_state_t *state)
 
 bool sdhc_send_cmd(sdhc_state_t *state, uint32_t cmd, uint32_t arg)
 {
-    uint32_t timeout = 0xFFFFFFFF;
+    uint32_t timeout = 0xFFFF;
```

Comment:
The timeout handling in this function is a bit error-prone. I wished it would be rather time-driven than driven by the core clock.

The timeout is only relevant for R1B responses (when the `MCI_RESP_BUSY` flag is set). All other responses must arrive within 64 clock cycles, which is handled by `SDHC_EISTR_CMDTEO` before the loop that checks the busy signal on DAT0 and generates this timeout.

Commands that normally use R1B are CMD7 (when selecting a card), CMD12, CMD20, CMD28, CMD29, CMD38, CMD39, CMD43, as well as CMD5 for SDIO and CMD6 for MMCs. However, of these commands, the driver uses this function only with CMD5 (to detect the card type), CMD7 to select the card during initialization, and CMD38 to erase sectors.

Furthermore, CMD7 can only lead to the busy signal on DAT0 if the card was previously deselected while being in the programming state and is selected again. Since CMD7 is only used during the card initialization to select the card once, this should not happen. So the only command for which this timeout has to be considered is CMD38 to erase a number of sectors. For CMD38, the timeout depends on the number of sectors erased. It is usually calculated as n * 250 ms.

If `UINT32_MAX` corresponds to about 10 min, `UINT16_MAX` should correspond to about 9 ms which could to be too small for erasing a number of sectors.

A better solution could be to return from this function and to use ACMD13 to wait for the card to leave the programming state calling `_wait_not_busy`.